### PR TITLE
Expand certificate support for LDAP authentication

### PR DESCRIPTION
**Pull request by JoshuaHassler**
- Allow user to specify Root CA to validate LDAP server cert against
- Allow user to specify client TLS cert/key for authentication with LDAP server

Proposed implementation for #134

### CODE REVIEW

File: LDAP-Auth/Config/PluginConfiguration.cs [hunk 1]
```diff
@@ -98,6 +101,21 @@ public PluginConfiguration()
...
+
+        /// <summary>
+        /// Gets or sets the ldap client cert path.
+        /// </summary>
+        public string LdapClientKey { get; set; }
+
+        /// <summary>
+        /// Gets or sets the ldap root CA path.
+        /// </summary>
+        public string LdapRootCa { get; set; }
```
Comment (crobibero) [inline LDAP-Auth/Config/PluginConfiguration.cs:117]:
```suggestion
        /// <summary>
        /// Gets or sets the ldap client cert path.
        /// </summary>
        public string LdapClientCertPath { get; set; }

        /// <summary>
        /// Gets or sets the ldap client cert path.
        /// </summary>
        public string LdapClientKeyPath { get; set; }

        /// <summary>
        /// Gets or sets the ldap root CA path.
        /// </summary>
        public string LdapRootCaPath { get; set; }
```

Make the properties descriptive of what they contain

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 2]
```diff
@@ -19,7 +22,7 @@ namespace Jellyfin.Plugin.LDAP_Auth
     /// </summary>
     public class LdapAuthenticationProviderPlugin : IAuthenticationProvider, IPasswordResetProvider
     {
-        private readonly ILogger<LdapAuthenticationProviderPlugin> _logger;
+        private static ILogger<LdapAuthenticationProviderPlugin> _logger;
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:25]: Revert this change and make `GetConnectionOptions` and `LdapClient_UserDefinedServerCertValidationDelegate` non-static

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 3]
```diff
@@ -210,11 +213,65 @@ public Task ChangePassword(User user, string newPassword)
...
+            SslPolicyErrors sslPolicyErrors)
+        {
+            if ((sslPolicyErrors & SslPolicyErrors.RemoteCertificateNameMismatch) > 0 ||
+                (sslPolicyErrors & SslPolicyErrors.RemoteCertificateNotAvailable) > 0)
+            {
+                _logger.LogWarning("Server certificate is not valid for remote domain");
+                return false;
+            }
+
+            X509Chain rootChain = new X509Chain
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:227]:
```suggestion
            using var rootChain = new X509Chain
```

`X509Chain` implements `IDisposable`. Please test after changing to make sure it still works

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 4]
```diff
@@ -210,11 +213,65 @@ public Task ChangePassword(User user, string newPassword)
...
+                    VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority |
+                                        X509VerificationFlags.IgnoreCertificateAuthorityRevocationUnknown |
+                                        X509VerificationFlags.IgnoreCtlNotTimeValid |
+                                        X509VerificationFlags.IgnoreCtlSignerRevocationUnknown |
+                                        X509VerificationFlags.IgnoreEndRevocationUnknown,
+                    TrustMode = X509ChainTrustMode.CustomRootTrust,
+                }
+            };
+            rootChain.ChainPolicy.CustomTrustStore.ImportFromPemFile(LdapPlugin.Instance.Configuration.LdapRootCa);
+            var result = rootChain.Build(new X509Certificate2(certificate));
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:240]:
```suggestion
            using var certificate2 = new X509Certificate2(certificate);
            var result = rootChain.Build(certificate2);
```

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 5]
```diff
@@ -373,10 +430,24 @@ private static LdapConnectionOptions GetConnectionOptions()
...
                 connectionOptions.ConfigureRemoteCertificateValidationCallback(LdapClient_UserDefinedServerCertValidationDelegate);
             }
 
+            if (!string.IsNullOrEmpty(configuration.LdapClientCert) && !string.IsNullOrEmpty(configuration.LdapClientKey))
+            {
+                var cert = X509Certificate2.CreateFromPemFile(configuration.LdapClientCert, configuration.LdapClientKey);
+                connectionOptions.ConfigureClientCertificates(new System.Collections.Generic.List<X509Certificate>()
+                    {
+                        cert,
+                    });
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:447]:
```suggestion
                connectionOptions.ConfigureClientCertificates(new[] { cert });
```

**Review by crobibero — CHANGES_REQUESTED** [review]

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs (hunk 2 above)
Comment (JoshuaHassler) [reply LDAP-Auth/LDAPAuthenticationProviderPlugin.cs]: This would also require a trickle-up of making `LdapConnectionOptions` non-static. Is that ok with you, or would you prefer I remove the extra logging from `LdapClient_UserDefinedServerCertValidationDelegate`?

Comment (crobibero) [reply LDAP-Auth/LDAPAuthenticationProviderPlugin.cs]: Yes, please make the function `GetConnectionOptions` non-static as originally stated

Comment (JoshuaHassler) [reply LDAP-Auth/LDAPAuthenticationProviderPlugin.cs]: Done. Thanks for the review!

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 6]
```diff
@@ -1,6 +1,9 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Net;
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:4]:
```suggestion
```
Unused

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 7]
```diff
@@ -208,13 +211,89 @@ public Task ChangePassword(User user, string newPassword)
...
+        /// <param name="sslPolicyErrors">Bitset of possible policy errors with the server's certificate.</param>
+        /// <returns>True if server cert is valid and trusted by the CA, otherwise false.</returns>
+        private bool LdapClient_UserDefinedServerCertValidationDelegate(
+            object sender,
+            X509Certificate certificate,
+            X509Chain chain,
+            SslPolicyErrors sslPolicyErrors)
+        {
+            if ((sslPolicyErrors & SslPolicyErrors.RemoteCertificateNameMismatch) > 0 ||
+                (sslPolicyErrors & SslPolicyErrors.RemoteCertificateNotAvailable) > 0)
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:229]:
```suggestion
            if (sslPolicyErrors.HasFlag(SslPolicyErrors.RemoteCertificateNameMismatch)
                || sslPolicyErrors.HasFlag(SslPolicyErrors.RemoteCertificateNotAvailable))
```

File: LDAP-Auth/LDAPAuthenticationProviderPlugin.cs [hunk 8]
```diff
@@ -208,13 +211,89 @@ public Task ChangePassword(User user, string newPassword)
...
+
+            using var rootChain = new X509Chain
+            {
+                ChainPolicy =
+                {
+                    VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority |
+                                        X509VerificationFlags.IgnoreCertificateAuthorityRevocationUnknown |
+                                        X509VerificationFlags.IgnoreCtlNotTimeValid |
+                                        X509VerificationFlags.IgnoreCtlSignerRevocationUnknown |
+                                        X509VerificationFlags.IgnoreEndRevocationUnknown,
```
Comment (crobibero) [inline LDAP-Auth/LDAPAuthenticationProviderPlugin.cs:243]:
```suggestion
                    VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority
                                        | X509VerificationFlags.IgnoreCertificateAuthorityRevocationUnknown
                                        | X509VerificationFlags.IgnoreCtlNotTimeValid
                                        | X509VerificationFlags.IgnoreCtlSignerRevocationUnknown
                                        | X509VerificationFlags.IgnoreEndRevocationUnknown,
```

Just moving the `|` to start of newline

**Review by crobibero — APPROVED** [review]
